src/target/riscv/riscv-013.c

```diff
@@ -3396,311 +3396,532 @@ static int write_memory_abstract(struct target *target, target_addr_t address,
 			return ERROR_FAIL;
-		while (get_field(abstractcs, DM_ABSTRACTCS_BUSY))
-			if (dmi_read(target, &abstractcs, DM_ABSTRACTCS) != ERROR_OK)
-				return ERROR_FAIL;
-		info->cmderr = get_field(abstractcs, DM_ABSTRACTCS_CMDERR);
+		index_on_target = (address_on_target - access.target_address) /
+			access.increment;
+	}
 
-		unsigned next_index;
-		unsigned ignore_last = 0;
-		switch (info->cmderr) {
-			case CMDERR_NONE:
-				LOG_DEBUG("successful (partial?) memory read");
-				next_index = index + reads;
-				break;
-			case CMDERR_BUSY:
-				LOG_DEBUG("memory read resulted in busy response");
+	/* According to the spec, if an abstract command fails, one can't make any
+	 * assumptions about dm_data registers, so all the values in the pipeline
+	 * are clobbered now and need to be reread.
+	 */
+	const uint32_t min_index_on_target = start_index + 2;
+	if (index_on_target < min_index_on_target) {
+		LOG_TARGET_ERROR(target, "Arithmetic does not work correctly on the target");
+		return ERROR_FAIL;
+	} else if (index_on_target == min_index_on_target) {
+		LOG_TARGET_DEBUG(target, "No forward progress");
+	}
+	const uint32_t next_index = (index_on_target - 2);
+	*elements_read = next_index - start_index;
+	LOG_TARGET_WARNING(target, "re-reading memory from addresses 0x%"
+			TARGET_PRIxADDR " and 0x%" TARGET_PRIxADDR,
+			access.target_address + access.increment * next_index,
+			access.target_address + access.increment * (next_index + 1));
+	return read_memory_progbuf_inner_startup(target, access.target_address,
+			access.increment, next_index);
+}
 
-				increase_ac_busy_delay(target);
-				riscv013_clear_abstract_error(target);
+/**
+ * This function attempts to restore the pipeline after a dmi busy.
+ */
+static int read_memory_progbuf_inner_on_dmi_busy(struct target *target,
+		uint32_t start_index, uint32_t next_start_index,
+		struct memory_access_info access)
+{
+	LOG_TARGET_DEBUG(target, "DMI_STATUS_BUSY encountered in batch. Memory read [%"
+			PRIu32 ", %" PRIu32 ")", start_index, next_start_index);
+	if (start_index == next_start_index)
+		LOG_TARGET_DEBUG(target, "No forward progress");
 
-				dmi_write(target, DM_ABSTRACTAUTO, 0);
+	if (dmi_write(target, DM_ABSTRACTAUTO, 0) != ERROR_OK)
+		return ERROR_FAIL;
+	return read_memory_progbuf_inner_startup(target, access.target_address,
+			access.increment, next_start_index);
+}
 
-				uint32_t dmi_data0, dmi_data1 = 0;
-				/* This is definitely a good version of the value that we
-				 * attempted to read when we discovered that the target was
-				 * busy. */
-				if (dmi_read(target, &dmi_data0, DM_DATA0) != ERROR_OK) {
-					riscv_batch_free(batch);
-					goto error;
-				}
-				if (size > 4 && dmi_read(target, &dmi_data1, DM_DATA1) != ERROR_OK) {
-					riscv_batch_free(batch);
-					goto error;
-				}
+/**
+ * This function extracts the data from the batch.
+ */
+static int read_memory_progbuf_inner_read_batch(struct target *target,
```

Comment:
You can rename this function to make it clearer what it is doing (that it only extracts data from a finished batch).

For example:

```suggestion
static int read_memory_progbuf_inner_read_data_from_batch(struct target *target,
```

or

```suggestion
static int read_memory_progbuf_inner_read_batch_data(struct target *target,
```

or 

```suggestion
static int read_memory_progbuf_inner_extract_batch_data(struct target *target,
```

... or something similar :)

Reply:
Addressed.